src/codec/dogstatsd.rs

```diff
@@ -0,0 +1,795 @@
+    };
+
+    // Optional Sections
+    let sections: Vec<&str> = substr(data, section_start..)?.split("|").collect();
+
+    for section in sections.iter() {
+        if section.starts_with('@') {
+            let sample_rate = &section[1..];
+            let sample_rate_float: f64 = sample_rate.parse()?;
+            m.insert("sample_rate".into(), Value::from(sample_rate_float));
+        } else if section.starts_with('#') {
+            let tags: Vec<&str> = section[1..].split(",").collect();
+            m.insert("tags".into(), Value::from(tags));
+        } else if section.starts_with('c') {
+            let container_id = &section[2..];
+            m.insert("container_id".into(), Value::from(container_id));
+        }
+    }
+
+    Ok(Value::from(m))
+}
+
+fn decode_event(data: &[u8]) -> Result<Value> {
+    let mut d = data.iter().enumerate();
+    let mut m = Object::with_capacity(13);
+    m.insert("dogstatsd_type".into(), Value::from("event"));
+    let section_start: usize;
+    let mut optional_sections = false;
+    let mut optional_text_idx = 0;
+
+    // Title/Text Lengths and Title
+    loop {
+        match d.next() {
+            Some((idx, b'|')) => {
+                let v: Vec<&str> = substr(data, 2..idx)?.split(":").collect();
+                let lens = v[0];
+                let len_vec: Vec<&str> = lens.split(",").collect();
+                let title_len: i32 = len_vec[0][1..].parse().unwrap();
+                let text_len: i32 = len_vec[1][0..len_vec[1].len() - 1].parse().unwrap();
+                let title = v[1];
+                m.insert("title_length".into(), Value::from(title_len));
+                m.insert("text_length".into(), Value::from(text_len));
+                m.insert("title".into(), Value::from(title));
+                section_start = idx + 1;
+                break;
+            }
+            Some(_) => (),
+            None => return Err(invalid()),
+        }
+    }
+
+    // Text
+    loop {
+        match d.next() {
+            Some((idx, _)) => {
+                let mut is_end = false;
+                let mut text_end_index = 0;
+                if idx == data.len() - 1 {
+                    is_end = true;
+                    text_end_index = idx;
+                } else if substr(data, idx..=idx)?.eq("|") {
+                    is_end = true;
+                    text_end_index = idx - 1;
+                    optional_sections = true;
+                    optional_text_idx = idx + 1;
+                }
+                if is_end && text_end_index > 0 {
+                    let text = substr(data, section_start..=text_end_index)?;
+                    m.insert("text".into(), Value::from(text));
+                    break;
+                }
+            }
+            None => return Err(invalid()),
+        }
+    }
+
+    // Optional Sections
+    if optional_sections {
+        let sections: Vec<&str> = substr(data, optional_text_idx..)?.split("|").collect();
```

Comment:
if you don't do `.collect()` here and keep it an iter, we can avoid an allocation for the `Vec`. Because below you just use it as an iterator, so no need to make it a `Vec` beforehand.

Sorry, if I may seem a little bit picky on those details.

Reply:
Neat! Thanks!